### DISCUSSION TITLE
Does this project work on Raspberry Pi ?

### DISCUSSION THREAD

**Discussion by Wil31** [daanzu/kaldi-active-grammar#49] [Q&A] [answered]
Hello, I do not manage to install the package kaldi-active-grammar on a Raspberry Pi. 
I get the following error:  Could not find a version that satisfies the requirement kaldi-active-grammar (from versions: none)
ERROR: No matching distribution found for kaldi-active-grammar
Can you please support me ?

**Comment by daanzu** [comment, marked as answer]
I have manually compiled and run it on a Raspberry Pi 4 with decent results, but I don't have automated builds for it yet. Hopefully if I find time.

**Reply by kris31** [reply]
Dear daanzu, I have the same error than Wil31.

Wil & I have superseded Intel Math Kernel by OpenBLAS as requested but the compilation fails.
For information, if we perform the following it does not work properly:
"""
sudo apt-get update && sudo apt-get upgrade
sudo apt-get install gfortran
python -m pip install -r requirements-build.txt
python setup.py bdist_wheel
cd /home/pi/kaldi-active-grammar-1.8.1/_skbuild/linux-aarch64-3.7/cmake-build/kaldi-prefix/src/kaldi/tools/extras
sudo ./install_openblas.sh
python setup.py bdist_wheel
""""
Either our process is wrong or something is missing!
Could you please provide an explanation on how to manually compile?
Best regards.
Chris

**Reply by kris31** [reply]
Here is what we get:
"""
An error occurred while building with CMake.
  Command:
    cmake --build . --target install --config Release --
  Source directory:
    /home/pi/Documents/kaldi-active-grammar-1.8.1
  Working directory:
    /home/pi/Documents/kaldi-active-grammar-1.8.1/_skbuild/linux-aarch64-3.7/cmake-build
Please see CMake's output for more information.
"""
...and CMakeOutput log looks like a hyeroglyph file for me!

**Reply by daanzu** [reply]
@kris31 Sorry, I haven't had a chance to look into this yet. I don't have my raspberry pi hooked up at the moment. For what it's worth, you should be able to find the important errors that are causing the build to fail: start at the bottom, and scroll up a bit until you start to see standard compilation commands being echoed, and you should see the error after them, followed by all of the build system spam.

**Reply by kris31** [reply]
Dear daanzu, thanks for your reply.
But honestly, we are three people trying to get kaldi-active-grammar running on RSP arm64 without any success!!!
As a second option, we have followed the manual process mentioned in your kaldi-fork:
Steps:
    (1)
    go to tools/  and follow INSTALL instructions there.

    (2)
    go to src/ and follow INSTALL instructions there.
....and then once all is done, we launch a "pip install 'dragonfly2[kaldi]' without any success.
Could you please explain us how do you proceed to get kaldi running on Raspberry Pi 4 ARM64 (image 2021-04-09) or could you provide us a more detailed process.
Thanks in advance.
Best regards.
Christophe R

**Reply by Wil31** [reply]
Hello @daanzu , thanks for your previous replies.
If it can help I can provide the following logs, issued while compiling on the Rasperry Pi 4 64bit. 
Currently, I manage to build and get the binaries in the directories "bin/", "gmmbin/", featbin/", "fstbin/", and "lm/", all subdirectories of "src/". But then, I don't know what to do with them in order to create a wheel or something we can install. (or a make install which does not exist)

[CMakeError.log](https://github.com/daanzu/kaldi-active-grammar/files/6319177/CMakeError.log)
[CMakeOutput (1).log](https://github.com/daanzu/kaldi-active-grammar/files/6319178/CMakeOutput.1.log)

My understanding about compilation is quite limited, but I'm willing to provide any information that could help more.

**Reply by daanzu** [reply]
@Wil31 First off, I would advise using the latest 2.1+ version, as it is already set up to use OpenBLAS. You will want to adjust the build settings on line for what ARM aarch64 needs: https://github.com/daanzu/kaldi-active-grammar/blob/master/CMakeLists.txt#L78

**Reply by Wil31** [reply]
Hello, 
Little update, we have managed to compile kaldi-active-grammar 2.1.0 to Raspberry Pi OS 64 bit beta running on the Raspberry Pi 4 4GB. And it runs with good performance. 
It was finally quite straightforward, after installing the required packages for compiling and running the setup to create the wheel.
Don't hesitate if you need more info.
